File: 15-3-sum/solution.ts
export function threeSum(nums: number[]): number[][] {
  function sorted2Sums(
    nums: number[],
    start: number,
    target: number,
  ): [number, number][] {
    const result: [number, number][] = [];
    const tracker = new Set<number>();
    let end = nums.length - 1;
    while (start < end) {
      const numStart = nums[start] as number;
      const numEnd = nums[end] as number;
      if (tracker.has(numStart)) {
        ++start;
        continue;
      }

      const needed = target - numStart;
      if (needed === numEnd) {
        result.push([numStart, numEnd]);
        tracker.add(numStart);
        continue;
      }

      if (needed < numEnd) {
        --end;
      } else {
        ++start;
      }
    }

    return result;
  }

  const result: number[][] = [];
  const visited = new Set<number>();
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length; ++i) {
    const iNum = nums[i] as number;
    if (iNum > 0) {
      break;
    }

    if (visited.has(iNum)) {
      continue;
    }

    visited.add(iNum);
    const combos = sorted2Sums(nums, i + 1, -iNum);
    for (const [jNum, kNum] of combos) {
      result.push([iNum, jNum, kNum]);
    }
  }

  return result;
}

File: 15-3-sum/solution.py
from typing import List


class Solution:
  def sorted_2_sums(self, nums: List[int], start: int, target: int) -> List[tuple[int, int]]:
    result: List[tuple[int, int]] = []
    tracker = set()
    end = len(nums) - 1
    while start < end:
      num_start, num_end = nums[start], nums[end]
      if num_start in tracker:
        start += 1
        continue

      needed = target - num_start
      if needed == num_end:
        result.append((num_start, num_end))
        tracker.add(num_start)
        continue

      if needed < num_end:
        end -= 1
      else:  # needed > num_end
        start += 1

    return result

  def threeSum(self, nums: List[int]) -> List[List[int]]:
    result = []
    visited = set()
    nums.sort()
    for i, i_num in enumerate(nums):
      if i_num > 0:
        break

      if i_num in visited:
        continue

      visited.add(i_num)
      combos = self.sorted_2_sums(nums, i + 1, -i_num)
      result.extend([[i_num, j_num, k_num] for j_num, k_num in combos])

    return result
